crates/agent/src/jobs.rs

```diff
@@ -174,9 +177,17 @@ where
     }
     .map_err(Error::Stderr);
 
-    let wait = child.wait().map_err(Error::Wait);
+    let status = child.wait().map_err(Error::Wait).await?;
 
-    let (_, _, _, status) = futures::try_join!(stdin, stdout, stderr, wait)?;
+    // If you don't await a future then it doesn't run, so the consequence
+    // of this should be to only stream std(out|err) to logs if there was an error,
+    // or if only_log_on_error wasn't set
+    if !only_log_on_error || (only_log_on_error && !status.success()) {
+        logs::write_line(logs_tx.clone(), name.to_string(), logs_token, &format!("Error while running command: {}", name)).await.unwrap();
+        futures::try_join!(stdin, stdout)?;
```

Comment:
- I think this could potentially cause a deadlock if the process outputs enough data to fill the stdout buffer. The child process may never exit if it's unable to write it's output, and nothing will be consuming that output until we try to join here.
- I don't see where the stderr future is ever polled

I think the solution here will require actively reading both stdout and stderr into growable in-memory buffers, and then writing them to the logs in a separate step if the command fails.

Reply:
Ah okay, that makes sense